### PR TITLE
disklimit: Allow more than 26 disks

### PR DESCRIPTION
**Pull request by jonathangold**
## Tips:

* commit message titles must be in the form:
```topic: Capitalized message with no trailing period```
or:
```topic, topic2: Capitalized message with no trailing period```

* please rebase your patch against current git master:
```
git checkout master
git pull origin master
git checkout your-feature
git rebase master
git push your-remote your-feature
hub pull-request	# or submit with the github web ui
```

* after a patch review, please ping @purpleidea so we know to re-review:
```
# make changes based on reviews...
git add -p		# add new changes
git commit --amend	# combine with existing commit
git push your-remote your-feature -f
# now ping @purpleidea in the github PR since it doesn't notify us automatically
```

### CODE REVIEW

File: resources/virt.go [hunk 1]
```diff
@@ -1009,7 +1009,7 @@ func (d *cdRomDevice) GetXML(idx int) string {
 	b += "<disk type='file' device='cdrom'>"
 	b += fmt.Sprintf("<driver name='qemu' type='%s'/>", d.Type)
 	b += fmt.Sprintf("<source file='%s'/>", source)
-	b += fmt.Sprintf("<target dev='hd%s' bus='ide'/>", (string)(idx+97)) // TODO: 26, 27... should be 'aa', 'ab'...
+	b += fmt.Sprintf("<target dev='vd%s' bus='virtio'/>", NumToAlpha(idx))
```
Comment (purpleidea) [inline resources/virt.go]: I like the virtio change, but it should be in a separate commit please.

File: resources/virt.go [hunk 2]
```diff
@@ -1195,3 +1195,15 @@ func expandHome(p string) (string, error) {
 
 	return p, nil
 }
+
+func NumToAlpha(idx int) string {
```
Comment (purpleidea) [inline resources/virt.go]: Can you make this private please? Eg `numToAlpha` instead?

File: resources/virt.go [hunk 3]
```diff
@@ -1195,3 +1195,15 @@ func expandHome(p string) (string, error) {
...
+	var result string
+	var mod int = idx % 26
+	var div int = idx / 26
+	if div > 0 {
+		result = NumToAlpha(div - 1) + string(rune(mod+97))
```
Comment (purpleidea) [inline resources/virt.go]: This can be simplified and inlined, eg

```
return NumToAlpha(div - 1) + string(rune(mod+97))
```

and same for below, dropping the else.

File: resources/virt.go [hunk 4]
```diff
@@ -1195,3 +1195,15 @@ func expandHome(p string) (string, error) {
...
+	} else {
+		result = string(rune(mod + 97))
```
Comment (purpleidea) [inline resources/virt.go]: Can we replace 97 with `int('a')` to make it more obvious?

File: resources/virt_test.go [hunk 5]
```diff
@@ -46,3 +46,22 @@ func TestExpandHome(t *testing.T) {
...
+
+func TestNumToAlpha(t *testing.T) {
+	var NumToAlphaTests = []struct {
+		number int
+		result string
+	}{
+	{0, "a"},
+	{25, "z"},
+	{26, "aa"},
+	{27, "ab"},
```
Comment (purpleidea) [inline resources/virt_test.go]: please add a test for "aaa" and "aab", and maybe something longer like "cool". Might as well show off now that you wrote this lovely function :)

**Comment by jonathangold** [comment]
Ok. will do. I'll ping you when the changes are made.

**Comment by purpleidea** [comment]
@jonathangold PS: The linter found two small issues to fix: https://travis-ci.org/purpleidea/mgmt/jobs/273062586#L1022

**Comment by purpleidea** [comment]
@jonathangold Patch looks perfect, two small fixes needed:

1) Commit message prefix should be `virt: Allow more than 26 disks` or `resources: virt: Allow more than 26 disks`

2) Can you break out the ide->virtio change into a separate patch please?

3) The linter found two small issues...

Thanks!

**Comment by jonathangold** [comment]
@purpleidea, done. Sorry for the duplicate commit messages.

**Comment by purpleidea** [comment]
I think all the changes look good, one small battle to fight with git:

In the branch there are about seven commits. How do we squish these down into a single one?

# from your branch
git rebase -i master

a text editor will open, you'll want to change the "pick" string to say "fixup" or "f" on each commit except the last one.

Rebase will destructively change your git commits, so be careful if you're not sure what this does.

The ide->virtio patch could be excluded from the rebase -i

After this git push origin yourbranch -f

If you're not sure about this process, ping and we can point you to a longer tutorial.

**Comment by jonathangold** [comment]
I rebased, but still see 4 commits. Not sure why.

**Comment by purpleidea** [comment]
@jonathangold If you type `git log` you should see what's present in your branch...

You might only have been able to rebase everything except for two if you accidentally did development on git master. You should always develop in a branch. If `git log master` shows Jonathan Gold anywhere, you need to remove those. I can tell you how in another message. Probably easy to back and forth on IRC instead of here.

**Comment by jonathangold** [comment]
Thanks. That was it. I accidentally made changes to master. Now there is just one commit.

**Comment by purpleidea** [comment]
@jonathangold It's true we're down to one commit, but that commit is missing most of the code...

**Comment by jonathangold** [comment]
fuuuuuuu. I'll fix it.

**Comment by purpleidea** [comment]
@jonathangold Wear your git battle scars proudly. We've all been there, and we learn from it.

**Comment by purpleidea** [comment]
@jonathangold I think you've made some mistakes using git rebase... As I mentioned it's dangerous. Here's the easy fix:

Copy the whole repository somewhere...

Create a new one with: git clone --recursive https://github.com/purpleidea/mgmt

In that folder, `git checkout -b feat/my-patch`

Copy over the code into to it. `git add -p` the relevant parts.

`git commit`

git push origin feat/my-patch

LMK

**Comment by roidelapluie** [comment]
Also you might learn about git reflog in this case.

**Comment by purpleidea** [comment]
@jonathangold You just added the fix, but clobbered your old code :P I've got a backup if you need...

**Comment by jonathangold** [comment]
Solved. Finally :dancing_men:

**Comment by purpleidea** [comment]
@jonathangold This looks promising!! Hooray :) Let's see what travis has to say... (15 min)

File: resources/virt_test.go [hunk 6]
```diff
@@ -46,3 +46,25 @@ func TestExpandHome(t *testing.T) {
...
+		{27, "ab"},
+		{702, "aaa"},
+		{703, "aab"},
+		{63269, "cool"},
+	}
+
+	for _, test := range numToAlphaTests {
+		actual := numToAlpha(test.number)
+		if actual != test.result {
+			t.Errorf("numToAlpha(%s): number %s, result %s", test.number, test.result, actual)
```
Comment (purpleidea) [inline resources/virt_test.go]: go vet found one small issue:

```
resources/virt_test.go:67: arg test.number for printf verb %s of wrong type: int
```

int's need a `%d` not a `%s`.

Easy to fix... git add -p, git commit --amend, git push .... -f

Comment (jonathangold) [reply resources/virt_test.go]: ok. That should do it. Thanks for the heads up.

**Comment by purpleidea** [comment]
@jonathangold Check again more carefully I think...

**Comment by jonathangold** [comment]
Ok. Now it passes.

**Comment by purpleidea** [comment]
@jonathangold Merged! w00t, congrats slaying all the git daemons, and welcome to the project!

Let us know when you're ready to work something next level, and please join us in IRC.
